p4-16/spec/P4-16-spec.adoc

```diff
@@ -4880,7 +4880,7 @@ f(y = ya, x = xa);  // match arguments by name in any order
 
 The calling convention is copy-in/copy-out (<<sec-calling-convention>>).  For generic functions the type arguments
 can be explicitly specified in the function call. The compiler only
-inserts implicit casts for direction `in` arguments to methods or
+inserts implicit casts for direction `in` or directionless arguments to methods or
 functions as described in <<sec-casts>>.  The types for all
```

Comment:
One more thing, the sentence says "to methods or functions", but there are two more cases that could be relevant:
- directionless arguments of actions with argument explicitly supplied in P4 (I think you have mentioned that for this case the casts are already allowed somewhere);
- arguments of extern and control/parser instances are are directionless arguments, and implicit casts could be applied here too (i.e. `Register<bit<32>, bit<3>>(8, 42)` -- the intializer type is presumably `bit<32>`, but one expects `int` argument will work).

So essentially I would suggest this should be allowed in any cases where a directionless argument can happen (except for directionless argument value coming from control plane, but there I can't see a way how one could want to cast it).